### ISSUE TITLE
[question] differences between constants and parameters

### ISSUE THREAD

**Issue by fpvandoorn** [leanprover/lean3#554]
When I write

``` lean
section
parameters {A : Type} (a : A) (...)
  (* stuff *)
end
```

My **intuition** of `A` and `a` is that as long as I stay inside the section, they are treated as constants i.e. treated exactly the same as

``` lean
constants {A : Type} (a : A) (...)
  (* stuff *)
```

Only when we end the `section`, magic happens, and the parameters are abstracted inside the definitions.

However, this intuition is false. In Lean they _are_ treated differently, as [multiple](https://github.com/leanprover/lean/issues/529#issuecomment-89869240) [issues](https://github.com/leanprover/lean/issues/530) indicate.

There are probably more differences. Can you give a (partial) list of differences between `parameters` (as long as you remain in the section) and `constants`?

**Comment by leodemoura** [comment]
Let me first explain how the parameters are implemented.
When we declare a definition `foo` in a section:
1- Collect all parameters used by `foo`. This includes the explicit and implicit (by transitivity) ones.
2- Update the type (and value) of `foo` will all new parameters. 
3- Elaborate
4- Send elaborated `foo` to the kernel
5- Create an alias `foo` ==> `foo As` where `As` are the parameters used by `foo`

So, the declaration is abstracted right away, and we use the alias to hide this fact from the user.
This simulation "leaks" sometimes (e.g., the pretty printer issue you found).

The more important question is: why do we use this approach?
In Lean 0.1, we used an approach similar to the one used in Coq. Parameters were added like axioms/constants in the environment, then in the end of the section all declarations are abstracted with respect to these axioms/constants. This approach is simpler, but it has problems.
In March of 2014, Georges pointed out several of them (at the time, we did not have google groups for Lean, and we discussed this issue in an email between Jeremy, Cody and me). 

1- The type of a declaration changes over time. This affects the ability of compiling theorems in parallel. The "end section" is like a synchronization point.

2- We have to re-type-check the section declarations in the "end section" (like in Lean 0.1), or the abstraction mechanism must be part of the trusted code base (like in Coq). Georges pointed out that he found bugs in this part of Coq.

3- We need to implement a mechanism for preserving meta-information (attributes, notation, etc). 
In Lean 0.1, suppose we have

``` lean
...
-- Environment E1
section 
...
-- Environment E2
end
-- Environment E3
...
```

Then, to obtain `E3`, we start from `E1` and add all declarations added in the section. That is, we get them from `E2`, abstract them, and send to the new environment we are building. This process must also copy some of the meta-information declared in the section.
In Coq, most of the meta-information is lost. See the following remark from Coq's manual: "Most commands, like Hint, Notation, option management, ... which appear inside a section are canceled when the section is closed."

4- Georges pointed out that inside of a section he frequently needs the abstract version of a declaration `foo`. So, the only way is to close the section. However, after he closes the section everything in the section is abstracted. He says this is really annoying. 

All the issues above are fixed by the approach used to implement parameters in Lean 0.2.
However our approach has two problems:

1- The simulation "leaks" sometimes. Hopefully, all leaks will be fixed one day, and this will not be an issue anymore.

2- I think it may be a performance bottleneck if we have hundreds of parameters. When we have a big section with hundreds or thousands of parameters, it seems to be much more efficient to treat them as constants, and only at the very end abstract everything. We have much more compact terms inside the section. Since we usually do not have hundreds of parameters, I don't think this is a big deal.

**Comment by fpvandoorn** [comment]
This is very helpful to have in one place. You've convinced me that the current implementation is indeed better, provided there are no "leaks". 

So, to summarize the differences between `parameter` (while in the section/context) and `constant` from a user perspective:
- With parameters you can access the abstracted version of `foo` using `namespace.foo` (or `_root_.foo` in top-level)
- Ignoring meta-information, everything else should be the same, right?

About meta-information, I know we have had discussions about this in the past (e.g. [here](https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/lean-discuss/nCn-5xTf13s/ebG5BWXnl9UJ)), so excuse me for bringing it up again. Let me start by summarizing the current implementation (please correct me where I'm mistaken):

Meta-information is either a `notation`, an `abbreviation` or an `attribute`. You can declare meta-information locally, which lasts until the end of the namespace (or until the end of the file when at top-level). All meta-information (local or not) declared in a context will be forgotten when the context is closed.

When parameters are around, a user could mean two things when giving meta-information: it applies for the abstracted version ("`namespace.foo`"), or the alias ("`foo`"). 
Regarding attributes, I think we currently only allow attributes for the abstracted version. 
For notation, we allow notation both inside sections and contexts, and they can refer to parameters. Since parameters are abstracted away at the end of a section/context, we only allow _local_ notation in sections (in context it doesn't matter: we forget all local and global notation anyway).

This is my knowledge of the current implementation. Please correct any mistakes. I vaguely recall that a section is a "nameless" namespace, but there seem to be differences. (EDIT: this was mentioned [here](https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/lean-discuss/7Tx5pDYVAEQ/cafRdtKv0coJ), but that seems to be outdated. Three current differences are how they treat local meta-information; whether you can declare parameters in them; and whether you can open namespaces in them)

To be blunt: I think the current implementation is messy. 
If you open a context, there is no way to preserve any meta-information for later (when you've closed the context). It seems that whatever you can do inside a context, you can do inside a section (by declaring all meta-information as `local`), and if you want to preserve at least one thing for later: you cannot use a context, you have to use a section.
Moveover: I think the fact that _local_ meta-information can persist past the end of a section is buggy. This example illustrates this:

``` lean
open nat
section
parameters (b : ℕ)
definition add_b (n : ℕ) := n + b
local postfix `%`:max := add_b
end
eval 5%
-- This is evaluated as `5 + b`, but there is no `b` in the context.
```

Also, it is slightly annoying that you cannot declare persistent notation which does _not_ contain a reference to parameters inside a section.

Ideally, I think we want to have something like the following (but please point out any inconveniences/problems with it):
- remove `contexts`
- delete all `local` meta-information declared in a section when ending the section.
- when declaring persistent notation inside a section, check whether the value contains a reference to a parameter (including definitions which depend a parameter). If it does, raise an error, otherwise accept it. Local notation is accepted if it contains parameters
- For attributes, persistent attributes can be applied to definitions with parameters, but then they are only applied to the _abstracted_ versions.
- Eventually (this can be implemented much later than the other points) it would be nice that for every attribute we have a `local attribute` which acts as an attribute for the alias (and hence would act the same way as if all `parameters` were replaced by `constants`). Then [local coercions](https://github.com/leanprover/lean/issues/525) and [local instances](https://github.com/leanprover/lean/issues/529#issuecomment-89869240) are special cases of this. This is probably an enormous amount of work to implement, because it is different for every attribute. It's probably not useful for every attribute (`local attribute foo [c-unfold n]` doesn't look very useful). Is there a full list of attributes somewhere?

Again: please point out any problems this change has (other than the obvious one which is that it is a lot of work to implement).

**Comment by leodemoura** [comment]
> Ignoring meta-information, everything else should be the same, right?

Another difference is that parameters cannot be universe polymorphic, but constants can.

> Meta-information is either a notation, an abbreviation or an attribute.

These are the main ones. We also have `calc_trans` and `tactic_hints`.
Moreover, "attributes" is a "umbrella" for many different modules: coercions, classes/instances,  reducible-hints, etc.

> This is my knowledge of the current implementation. Please correct any mistakes. I vaguely recall that a section is a "nameless" namespace, but there seem to be differences.

Yes, the meta-objects persist sections. This is useful when we are using a section with parameters and many declarations and attributes, and we don't want to reset all attributes after we close the section.

> To be blunt: I think the current implementation is messy. 

The section/context mechanism has been revised so many times that it is hard to have a clean implementation. Each modification affects many modules (i.e., any module that creates meta-objects).

> It seems that whatever you can do inside a context, you can do inside a section (by declaring all meta-information as local), and if you want to preserve at least one thing for later: you cannot use a context, you have to use a section.

The contexts may look redundant **now**, but this was not always the case. Features such as local attributes are relatively new. In Lean 0.2, parameters were not initially supported in sections.

> Moveover: I think the fact that local meta-information can persist past the end of a section is buggy.

We can fix this.

> This is probably an enormous amount of work to implement, because it is different for every attribute.

Yes, it a lot of work, and there are potential problems. Example: to be able to compile theorems in parallel, we must be able to infer their types before we elaborate them. So, before we start the elaboration process, we must be able to decide all parameter a theorem depends on. If we allow local instances with parameters, then we are in trouble. We will not be able to decide the exact list of parameters until we elaborate the whole theorem. If the type class resolution picks an instance containing an extra parameter, we will produce an invalid term. 

Another problem: a section containing local coercions (with parameters) and persistent coercions (without parameters). The coercion module computes the transitive closure of the coercion graph. So, we have to fix the coercion module to take that into account. That is, given

``` lean
    f [coercion] : A -> B
 local g [coercion] : B -> C 
```

The derived coercion `g o f`  must be local. Another approach is to delete this module, and implement a new one where we do **not** maintain the transitive closure. 

It is hard to enumerate all problems, I would have to go over all the modules that create attributes and analyze them case by case.

I agree that `context`s look very redundant now. However, unless I have help, it is a lot of work to make the kind of change you suggest.

**Comment by leodemoura** [comment]
Here is a compromise, we can try to implement your suggestion in stages.
Let's first try to remove the `context`s. Then, we try to remove the restriction on notation. 
Later, we try to add a new coercion module which supports local and non-local coercions.
We would also be useful to have a definitive solution for parameters/variables (issue #532)

**Comment by fpvandoorn** [comment]
Thanks for the corrections.

> Yes, the meta-objects persist sections. This is useful when we are using a section with parameters and many declarations and attributes, and we don't want to reset all attributes after we close the section.

Do we want local meta-information to persist?

> The section/context mechanism has been revised so many times that it is hard to have a clean implementation. Each modification affects many modules (i.e., any module that creates meta-objects).

Yes, I understand. It was not meant as criticism, but I think it will be confusing for a new user.

> The contexts may look redundant now, but this was not always the case. Features such as local attributes are relatively new. In Lean 0.2, parameters were not initially supported in sections.

I understand. I think it is a good thing that we allow parameters in sections now, and that as a consequence we can remove contexts: it is one fewer feature a user has to learn.

> Yes, it a lot of work, and there are potential problems.

I see. I can indeed imagine that each attribute has its own problems. It's indeed much harder than "just mimic what happens when we would have used constants". 

Side note about the local instances: we could adopt the rule that a local instance `foo` which depends on parameter `A` is only available if we use `include A`. This is a similar to the rule that a variable is only available for type class resolution if we `include` it. 

> Here is a compromise, we can try to implement your suggestion in stages.

Yes, that is a good plan. I agree with your stages. Considering local attributes, local coercions seem to be the most useful.

**Comment by leodemoura** [comment]
> Side note about the local instances: we could adopt the rule that a local instance foo which depends on parameter A is only available if we use include A. This is a similar to the rule that a variable is only available for type class resolution if we include it.

We have an `omit` command :-(
So, the user can do the following `include A`, define instance, `omit A`.

**Comment by leodemoura** [comment]
> Do we want local meta-information to persist?

No, we don't. This is a bug in the section mechanism. We wanted for attributes in sections to persist, then we added local attributes (and a bug: local attributes in sections persist in the outer scope but not in the .olean file).

**Comment by fpvandoorn** [comment]
> We have an omit command :-(
> So, the user can do the following include A, define instance, omit A.

I mean that whenever parameter `A` is included, then the instance if available. If you later `omit A`, then the instance becomes unavailable. This is similar for variables/parameters which are instances themselves:

``` lean
variable [H : is_inhabited A]
include H
-- H is available for type class resolution
omit H
-- H is unavailable
```

So 

``` lean
parameter (A : T)
definition foo : _ := (... A ...)
local attribute foo [instance]
include A
-- foo is available for type class resolution
omit A
-- foo is unavailable
```

**Comment by leodemoura** [comment]
> I mean that whenever parameter A is included, then the instance if available. If you later omit A, then the instance becomes unavailable.

This is non modular. The `include A` command affects a completely different module: type class resolution. The type class resolution module is unaware of the existence of parameters/include/omit/etc.
Here is a potential "spaghetti" implementation:
- Type class resolution module has a flag (enable/disable) associated with each instance.
- Frontend maintains a mapping `M` from parameters to instances
- Whenever we declare an instance, it is enabled iff all parameters occurring in the instance are "included"
- Whenever we include/omit, we use `M` to visit the instances that may be affected, and we disable/enable them.

**Comment by leodemoura** [comment]
Here is a update:

1- Fixed the `section` local metadata issue. Now, the following example produces the expected behavior

``` lean
open nat
section
parameters (b : ℕ)
definition add_b (n : ℕ) := n + b
local postfix `%`:max := add_b
end
eval 5%
```

2- Removed all occurrences of `context` command from the standard and hott libraries, and test suite.

3- Fixed bugs in the standard library. Some files relied on the bug in the `section` command. Example:

``` lean
  section
    open eq.ops
    ...
  end
  -- eq.ops notation is not available anymore
  check Ht ▸ He --  ▸ should not be available
```

4- Notation defined in a context always overrides global notation (i.e., no overloading). So, to have a similar capability in sections, I added the following modification: local notation in a section overrides global one. 

5- I fixed other bugs in the section with parameters. The commits submitted today describe them.

One gotcha: in contexts, `universe u` is always treated as `universe variable u`. We have to remember that when we move code using a context into code using a section. Another option is to also treat `universe u` in a section as `universe variable u`. 

I think we are ready to delete the `context` command.
@avigad @javra: any objections?
I agree with Floris, the `context` command is now redundant. Less is always better :-)

**Comment by fpvandoorn** [comment]
> This is non modular.

I see. That is indeed a difference. Type class inference can ignore variables which are not `included` because they are just not in the local context. That is different with local instances.

> Here is a update:

Great! I see that you have some fancy tests with nested sections :-)

> Another option is to also treat universe u in a section as universe variable u. 

I have no opinion here. I have never used `universe u`. Is `universe u` the version of `constant t` for universes? And `universe variable u` looks like a mix between `parameter` and `variable`: it is abstracted in definitions/theorems, but not in variables/parameters. Ok, maybe I have a small opinion: I think we can allow `universe u` in a section (and not interpret it as `universe variable u`). But if it is convenient to it in an other way, then that's also fine.

**Comment by leodemoura** [comment]
>  Is `universe u` the version of `constant t` for universes?

Yes.

**Comment by avigad** [comment]
Let me know if / when I should remove the discussion of contexts from the tutorial.

**Comment by leodemoura** [comment]
> Let me know if / when I should remove the discussion of contexts from the tutorial.

@avigad So, I'm assuming you have no objections.

**Comment by leodemoura** [comment]
BTW, after we remove `context`. Then, all but the following item in Floris' proposal will be covered:

> Eventually (this can be implemented much later than the other points) it would be nice that for every attribute we have a local attribute which acts as an attribute for the alias (and hence would act the same way as if all parameters were replaced by constants). Then local coercions and local instances are special cases of this. This is probably an enormous amount of work to implement, because it is different for every attribute. It's probably not useful for every attribute (local attribute foo [c-unfold n] doesn't look very useful). Is there a full list of attributes somewhere?

Floris is right. This one is an enormous amount of work.
Right now, all attributes local or not are for the _abstracted_ version. 
I think it too much work for a relatively modest improvement in usability. 
As I told with Jeremy yesterday, I want to focus all resources on better automation.

**Comment by fpvandoorn** [comment]
> I want to focus all resources on better automation. 

I agree that that is more important.

**Comment by avigad** [comment]
Here is one way I have used contexts in the past: I don't want to overload say, "<", except for one proof, so I do

``` lean
context temp
  open nat
  theorem foo : ...
end temp
```

Will this work with sections?

Another use was this: sometimes I want to open algebra (with all the classes and coercions) only temporarily. Will that work with sections?

**Comment by leodemoura** [comment]
> Will this work with sections?
> Will that work with sections?

After yesterday fixes, both work with sections.

**Comment by avigad** [comment]
Then I am happy to say goodbye to contexts. 

With "local notation", I see no need for them, and as Floris points out they are less flexible than sections. This will make things much easier to explain to users.

**Comment by leodemoura** [comment]
I removed the `context` command.
Thus, I will mark this issue as postponed and close it.
For the record, the missing feature is:

> Eventually (this can be implemented much later than the other points) it would be nice that for every attribute we have a local attribute which acts as an attribute for the alias (and hence would act the same way as if all parameters were replaced by constants). Then local coercions and local instances are special cases of this. This is probably an enormous amount of work to implement, because it is different for every attribute. It's probably not useful for every attribute (local attribute foo [c-unfold n] doesn't look very useful). Is there a full list of attributes somewhere?
